### DISCUSSION TITLE
Module Creator Doesn't Complete

### DISCUSSION THREAD

**Discussion by pauljanisko** [oqtane/oqtane.framework#884] [Q&A] [answered]
I was very excited to learn about Oqtane from Shaun's .netConf 2020 talk, and would like to use it on the next project starting next week.  I downloaded it from git, ran it in VS 2019 Community, installed db in setup, and created a common module by following the YouTube video series.  I'm now trying to create a Developer Module using the Module Creator.

I enter in the Owner Name/Module Name/Desc/External Template, location was good, then clicked Create Module.  It created most of the files in the directory, but my browser shows a ERR_CONNECTION_RESET partway through.  I opened the .sln file and see that the Add.razor and View.razor files are missing from the Client folder, as well as several in the root of the parent folder (only .sln file exists).  

I have turned off my firewall, do not use a proxy.  I use Windows 10 and it's up to date.  I have the latest VS 2019, installed latest .net 5 SDK and hosting exe's, installed latest SQL 2019.  I have tried (IISExpress) in debug mode, started without debug, and chose to run under Oqtane.Server to see the command prompt.  I tried with the Brave browser and Chrome (both up to date).  During the module creation process, the server just shuts down.  

I have seen the links to the module template files to download, but I would like to use this model since it's fast and convenient.  It's likely operator error or environment, but if I ran into it while following the videos, someone else must also be running into it.  Any help would be appreciated.

Thank you.

**Comment by mikecasas** [comment]
Did you try creating an internal or external module?

**Reply by pauljanisko** [reply]
I used External.  When I tried Internal, it also exited with Code 0, but it never made a directory or any files.  When External, the parent directory was made, and project directories were made with 'most' files.

Debug w/cmd window:
https://www.dropbox.com/s/9pu0wp50coxj4bh/2020-11-14_16-17-08.png?dl=0 

When I view the page (after "Internal") install:
https://www.dropbox.com/s/qedseb8vig1779v/2020-11-14_16-31-50.png?dl=0

When I view the page (after "External") install:
https://www.dropbox.com/s/6u6yebas6yjdmqp/2020-11-14_16-36-26.png?dl=0

Thank you for replying to this mikecasas.

**Reply by mikecasas** [reply]
I was able to create a new module. I'm not sure why you are getting the error message. Do you have the new .NET 5 downloaded.

**Reply by mikecasas** [reply]
sorry I see in your other post you did install .net 5. Do you have your code on github?

**Reply by jimspillane** [reply]
@mikecasas did you compile the new module before restarting Oqtane?  I see the same [red bar error](https://www.dropbox.com/s/6u6yebas6yjdmqp/2020-11-14_16-36-26.png?dl=0) that @pauljanisko is receiving until I compile the new module.

**Reply by mikecasas** [reply]
Yes, I always compile it. 
When you run the module creator, it doesn't put the files in the Oqtane Server's bin folder.

**Reply by pauljanisko** [reply]
I tried using the oqtane.org download link and the github repo download.  Both today, 11/14/2020.  Same result both times.  I didn't change anything except making a "hello world" common module, in line with the YouTube videos.  

I receive the red banner, but I have also compiled the 'partial' module that was scaffold out and received many lines of text in oqtane in the module area with lists of paths.  It was trying, but didn't work.  

It feels like a local environment issue, which I hate to waste the time of others, but if I ran into it, I am hoping others have ran into it too so it helps them and me.  I'm impressed how smooth everything seems to work and excited to utilize this as a baseline for future projects.

**Reply by mikecasas** [reply]
@pauljanisko why do you call it a partial module.

**Reply by pauljanisko** [reply]
Sorry for blending words that are confusing to dev..  I'm talking human, not OOP.   I meant 'partial' because what was auto-generated in the module directory was not complete, as compared to the YouTube video.  There were files missing.  I changed no code.  Another reason for "partial" is because it wasn't complete enough to render as it was supposed to with text from the index.razor file.  Oqtane was confused and listed paths and things instead of reading the compiled solution.

Here is the "Client" folder of the output from External module creation:  https://www.dropbox.com/s/h20a4hxirdkyzel/2020-11-14_19-36-38.png?dl=0

**Reply by jimspillane** [reply]
There doesn't look like there are any missing files in that folder. The external module template files are located here:
https://github.com/oqtane/oqtane.framework/tree/master/Oqtane.Server/wwwroot/Modules/Templates/External
You can compare them to the files in your module.

Try opening the .sln in the parent of the _Client_ folder and compile the solution.  As @mikecasas said above, compiling it will add the module .dll to the Oqtane Server's bin folder.

**Reply by pauljanisko** [reply]
Thanks Jim.  I think this may be my best path forward.  

I get an error when creating it.  There are visibly files missing if you look in more detail:  https://www.dropbox.com/s/b50sbiw6buc03mc/2020-11-14_19-51-04.png?dl=0

This is an exact screenshot of the YouTube video of what should be vs what is.  

There are two files missing.  The module creation doesn't complete and files are missing.  I promise.  👍 

I did compile the module project.  I didn't verify the dll was placed into the oqtane bin, however the error message showed it knew about it but couldn't render it.

**Reply by mikecasas** [reply]
You can't compare that project to the module creator. The module creator scaffolds a starter template. The DNF project you reference has additional pages/components added for more functionality.

**Reply by pauljanisko** [reply]
Well, that's possible.  I watched video 1, and then video 2.  The files listed were from these steps.  No files were added and the module I made didn't render in oqtane.  

When you create a new module, do you have add.razor and view.razor in the client project?  I also mentioned that there are zero other files in the root of this module except the .sln file.

Video series I followed:  https://www.youtube.com/watch?v=JPfUZPlRRCE&list=PLYhXmd7yV0elLNLfQwZBUlM7ZSMYPTZ_f&index=1

**Comment by mikecasas** [comment]
No there is no Add and View razor file. Just Index and Edit razor files. The Edit component can handle Adds and Edits though. 

The root folder should just have the solution (*.sln file) and the four project folders. Client, Package, Server and Shared.

**Reply by mikecasas** [reply]
This is the basic file structure for a new module from the module creator.
![Main](https://user-images.githubusercontent.com/11261164/99162034-929bdb80-26c6-11eb-97cd-fdc7cf79e0bf.PNG)

**Comment by pauljanisko** [comment, marked as answer]
I've recorded a video of the full experience here:  https://www.dropbox.com/s/kwekuxh2d5drnpe/2020-11-15_08-51-42.mp4?dl=0

**Reply by jimspillane** [reply]
@pauljanisko after creating the module, try opening the _D:\_dev\QuoteBot.Quotes\QuoteBot.Quotes.sln_ and build the solution before reopening Oqtane.

**Reply by pauljanisko** [reply]
@jimspillane Thank you for your reply.  I have tried that as well.  Here is a recording of it:  https://www.dropbox.com/s/kwq2rge7ymj5kbd/2020-11-15_09-04-13.mp4?dl=0

**Reply by jimspillane** [reply]
That looks like a database error being thrown. Check to make sure the _QuoteBotQuotes_ table was created.  This should have happened when the module ran for the first time.

![image](https://user-images.githubusercontent.com/15007083/99191790-f82aaf00-273c-11eb-87c8-831d3dcf2e41.png)



The red bar message is from line 91 in the Index.razor file in the module.

![image](https://user-images.githubusercontent.com/15007083/99192010-6754d300-273e-11eb-87f8-9488df742d54.png)

**Reply by mikecasas** [reply]
@pauljanisko Try deleting everything in the Oqtane Server bin folder and start the process all over. After deleteing the bin folder, build oqtane, then build your external module.

Also check the database that your table (quotes??) was actually created.

**Reply by pauljanisko** [reply]
The module is rendering and working, however the error screen still displays after I click Create Module.  I rebuilt the module by opening the .sln file and rebuilding.  I then rebuilt Oqtane and started without debugging.  I removed the helpful paths text from the module index.razor, rebuilt the solution, then rebuilt Oqtane and started without debugging and have the screenshot below of QuoteBot Module Working.  

I started from scratch using my local SQLExpress (vs Oqtane.Server/Data/database mdf file), and then followed the above approach as suggested (and required) in the responses from this post.  Rebuilding the module and oqtane now shows this module working, even though I receive an ERR_CONNECTION_RESET after clicking "Create Module" button.  

Table Generated:
https://www.dropbox.com/s/5dfrh74kha08v39/2020-11-15_10-57-11.png?dl=0

QuoteBot Module Working:
https://www.dropbox.com/s/o6x81ucwrfdj4wl/2020-11-15_10-55-17.png?dl=0

I will forget about the error at this point and continue learning how to utilize the custom module and how I can build on top of the Oqtane framework.  I really appreciate the help and fast responses to this issue.  It gives me confidence moving forward with Oqtane, that the community is so supportive.

**Reply by jimspillane** [reply]
The ERR_CONNECTION_RESET behavior is expected.  Module creation [restarts the application](https://github.com/oqtane/oqtane.framework/blob/d9f9e734809b2d102fd3807af329b97d511bec25/Oqtane.Server/Controllers/ModuleDefinitionController.cs#L209) and there is no graceful way to reconnect at this time.

**Reply by mikecasas** [reply]
@pauljanisko, like @jimspillane said, the "error" you mention is just the app restarting. So you are good then? If so please mark the thread as answered.

**Reply by pauljanisko** [reply]
Yes, good to go.  Thanks guys!

**Comment by sbwalker** [comment]
The reason why the framework forces a server restart is because Oqtane relies on logic which executes during application startup to discover modules/themes and load them into the AppDomain - so if the application was not restarted, the module which was created using the Module Creator would never be activated. 

One thing I have noticed is that when people see the ERR_CONNECTION_RESET behavior they immediately assume that something went wrong. This behavior only happens when running on Blazor Server - and @jimspillane is correct that there is so far no technical solution to allow the application to recover when you force a server restart. 

One question I have is if it may be a better user experience in those areas of the framework which require a restart ( ie. module/theme creation, installation, deletion ) to not force the restart at that moment, but instead include a message to the user that the application needs to be restarted for the changes to take effect. We could include an option in one of the Admin UIs to allow a host user to manually force a restart the application. The benefit of this approach is that it puts the user in control of when the restart happens. It also consolidates restarts to a single deliberate area of the framework. And in the case where you are running Oqtane from Visual Studio you do not even need the manual option - you simply need to stop your debugging session and then run the app again. The downside is that if a user does not read the message displayed in the UI or understand that a restart is required, they may assume the framework is broken when their action does not provide their expected result immediately. 

Thoughts?

**Reply by jimspillane** [reply]
I like the idea of adding a Restart option in the AdminUI.  After successfully performing an action that requires a restart, a message can be returned to the user indicating "_...a restart is needed..._" and provide a link to the Admin page.

**Reply by pauljanisko** [reply]
In my case above, the application never came back.  A simple piece of text before I clicked "Create Module" that had a note stating the application will need to be restarted (link to admin page or just let user restart) and your browser may not come back.  Go ahead and build the module .sln and the Oqtane .sln and the module will be available in Oqtane.  

I was following the videos on YouTube exactly for setup, and this screen was not in the demo and the files in the demo didn't match the files generated on my machine.  I understand why now, but it took some time and help from this community.  I appreciate everyone's help and the level of activity surrounding Oqtane!

**Reply by sbwalker** [reply]
@pauljanisko Currently the Module Creator displays the following message at the top of the module "Please Note That Once You Select The Create Module Button The Application Must Restart In Order To Complete The Process. If You Create An External Module You Will Need To Compile The Source Code And Then Restart Your Site In Order To Make It Functional." I am guessing that you interpreted this to mean that the application will restart but will remain functional?

**Reply by pauljanisko** [reply]
@sbwalker I see that BIG PURPLE message in my screenshots, but honestly I was following along the YouTube video and didn't pay close (any) attention.  I fill out 5 fields like the video, and click the button.  I guess I was just eager to get into Visual Studio where I'm more comfortable.  Understanding what it's doing behind the scenes, it makes complete sense now that it would have to restart the application.  The message could have said, "By clicking this button, all puppies will be attacked by cats." and I wouldn't have noticed it.  I'm used to building things myself and didn't realize Oqtane did so much.  This wasn't just another form to fill out.  My bad.  

This is the point where you think, "here's your sign" (and "I hope he reads it").

**Reply by sbwalker** [reply]
No worries... personally I am happy you chose to give Oqtane a try, you used the video tutorials, the community helped you with your issue, and there is a suggested approach on how to improve the user experience
